src/libmongoc/src/mongoc/mongoc-crypt.c

```diff
@@ -576,37 +581,158 @@ _state_need_kms (_state_machine_t *state_machine, bson_error_t *error)
+   if (!bson_init_static (&subdoc, dataptr, datalen)) {
+      // Invalid "aws" document? Should be validated by libmongocrypt
+      return false;
+   }
+
+   if (bson_empty (&subdoc)) {
+      // "aws" is present and is an empty subdocument, which means that the user
+      // requests that the AWS credentials be loaded on-demand from the
+      // environment.
+      return true;
+   } else {
+      // "aws" is present and is non-empty, which means that the user has
+      // already provided credentials for AWS.
+      return false;
+   }
+}
+
+/**
+ * @brief Attempt to load AWS credentials from the environment and insert them
+ * into the given kmsProviders bson document on the "aws" property.
+ *
+ * @param out A kmsProviders object to update
+ * @param error An error-out parameter
+ * @return true If there was no error. We MAY or MAY NOT have successfully added
+ * the "aws" property.
+ * @return false If there was an error while updating the BSON data or obtaining
+ * credentials.
+ */
+static bool
+_try_add_aws_from_env (bson_t *out, bson_error_t *error)
+{
+   // Attempt to obtain AWS credentials from the environment.
+   _mongoc_aws_credentials_t got_creds;
+   if (!_mongoc_aws_credentials_obtain (NULL, &got_creds, error)) {
+      // Failed to get credentials. This isn't necessarily an error.
+      return error->code == 0;
+   }
+
+   // We have the credentials. Give them back to libmongocrypt
+
+   bson_t aws = {0};
+   // Begin the new "aws" subdoc
+   if (!BSON_APPEND_DOCUMENT_BEGIN (out, "aws", &aws)) {
+      return false;
+   }
+   // Add the accessKeyId and the secretAccessKey
+   if (!(BSON_APPEND_UTF8 (&aws, "accessKeyId", got_creds.access_key_id) &&
+         BSON_APPEND_UTF8 (
+            &aws, "secretAccessKey", got_creds.secret_access_key))) {
+      return false;
+   }
+   // We may have a sessionToken to add as well
+   if (got_creds.session_token) {
+      if (!BSON_APPEND_UTF8 (&aws, "sessionToken", got_creds.session_token)) {
+         return false;
+      }
+   }
+
+   // Finish the subdoc
+   if (!bson_append_document_end (out, &aws)) {
+      return false;
+   }
+
+   // Good!
+   return true;
+}
+
 static bool
 _state_need_kms_credentials (_state_machine_t *sm, bson_error_t *error)
 {
    bson_t creds = BSON_INITIALIZER;
-   BSON_ASSERT (sm->crypt->creds_cb.fn);
    const bson_t empty = BSON_INITIALIZER;
+   bool okay = false;
+
+   if (sm->crypt->creds_cb.fn) {
+      // We have a user-provided credentials callback. Try it.
+      const bool okay1 = sm->crypt->creds_cb.fn (
+         sm->crypt->creds_cb.userdata, &empty, &creds, error);
+      if (!okay1) {
+         if (!error->code) {
+            bson_set_error (error,
+                            MONGOC_ERROR_CLIENT_SIDE_ENCRYPTION,
+                            MONGOC_ERROR_CLIENT_INVALID_ENCRYPTION_ARG,
+                            "Unknown error from user-provided callback for "
+                            "on-demand KMS credentials");
+         }
```

Comment:
```suggestion
         if (!error->code) {
            bson_set_error (
               error,
               MONGOC_ERROR_CLIENT_SIDE_ENCRYPTION,
               MONGOC_ERROR_CLIENT_INVALID_ENCRYPTION_ARG,
               "user-provided callback for on-demand credentials failed");
         }
```
Suggest avoiding stating "unknown error"; user-provided callback may report additional details as desired.